### PR TITLE
[5.x] Udt 134933 s3 fileserver problems

### PR DESCRIPTION
**Pull request by tdrwenski**
-Add writeToStream method for MFile
-Add writeToStream method for a byte range of MFile
-Add tests

### CODE REVIEW

File: cdm/zarr/src/test/java/thredds/inventory/zarr/TestMFileZip.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package thredds.inventory.zarr;
```
Comment (tdrwenski) [inline cdm/zarr/src/test/java/thredds/inventory/zarr/TestMFileZip.java:1]: I just realized the `MFileZip:MFileProvider` doesn't get loaded by TDS (so TDS can only have `MFileS3` or `MFileOS` implementations returned from `MFiles.create`). So I'm not sure if these tests are very useful/important-- maybe I should remove them.

File: cdm/core/src/test/java/thredds/filesystem/TestMFileOS.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    final int endPosition = Math.min(offset + maxBytes, expectedSize);
+
+    assertThat(outputStream.size()).isEqualTo(Math.max(0, endPosition - startPosition));
+
+    final byte[] partialFile = Arrays.copyOfRange(Files.readAllBytes(file.toPath()), startPosition, endPosition);
+    assertThat(outputStream.toByteArray()).isEqualTo(partialFile);
+  }
+
+  private File createTemporaryFile(int size) throws IOException {
+    final File tempFile = File.createTempFile("TestMFileOS-", ".tmp");
```
Comment (haileyajohnson) [inline cdm/core/src/test/java/thredds/filesystem/TestMFileOS.java]: Would you mind changing this to use junit,rule.TemporaryFolder? Any files created in that get cleaned up automatically at the end of the test method, e.g.
```
  @ClassRule
  public static final TemporaryFolder tempFolder = new TemporaryFolder();
  File file = tempFolder.newFile();
```

File: cdm/core/src/main/java/thredds/inventory/CollectionManagerCatalog.java [hunk 3]
```diff
@@ -146,6 +147,16 @@ public Object getAuxInfo() {
...
+    @Override
+    public void writeToStream(OutputStream outputStream) throws IOException {
+      throw new IOException("Writing MFileRemote not implemented. Filename: " + getName());
+    }
+
+    @Override
+    public void writeToStream(OutputStream outputStream, long offset, long maxBytes) throws IOException {
+      throw new IOException("Writing MFileRemote not implemented. Filename: " + getName());
+    }
   }
```
Comment (haileyajohnson) [inline cdm/core/src/main/java/thredds/inventory/CollectionManagerCatalog.java:160]: I think this has the potential to break things down to line, but I also think that's a "down the line" problem that we can ignore for now.

File: cdm/zarr/src/test/java/thredds/inventory/zarr/TestMFileZip.java (hunk 1 above)
Comment (haileyajohnson) [reply cdm/zarr/src/test/java/thredds/inventory/zarr/TestMFileZip.java:1]: I think you should leave them, we're going to have to add MFileZip to the TDS if we want to support Zarr zip stores, which we probably do.

**Review by haileyajohnson — CHANGES_REQUESTED** [review]

File: cdm/core/src/main/java/thredds/inventory/CollectionManagerCatalog.java (hunk 3 above)
Comment (tdrwenski) [reply cdm/core/src/main/java/thredds/inventory/CollectionManagerCatalog.java:160]: I could also implement these, it is probably not very hard. But not necessary for this bug fix so I could do it in a separate PR

Comment (haileyajohnson) [reply cdm/core/src/main/java/thredds/inventory/CollectionManagerCatalog.java:160]: Yeah, up to you, but I think it's fine to wait and see how this first pass goes over in the TDS

File: cdm/core/src/test/java/thredds/filesystem/TestMFileOS.java (hunk 2 above)
Comment (tdrwenski) [reply cdm/core/src/test/java/thredds/filesystem/TestMFileOS.java]: Fixed in 93080b86dc

**Review by haileyajohnson — APPROVED** [review]

**Pull request merged**